File: core/sandbox.py
"""Sandbox — unrestricted workspace where ADAM can write and run anything.

This is ADAM's personal computer within the computer. It can:
- Write and run any code (Python, shell scripts, services)
- Start long-running processes (sub-agents, servers, daemons)
- Install packages
- Create entire applications
- No restrictions — if ADAM can imagine it, it can build and run it here.
"""

import os
import subprocess
import signal
import json
import time


SANDBOX_DIR = "/app/sandbox"
PROCESSES_FILE = "/app/sandbox/.processes.json"


def init():
    os.makedirs(SANDBOX_DIR, exist_ok=True)
    os.makedirs(os.path.join(SANDBOX_DIR, "projects"), exist_ok=True)
    os.makedirs(os.path.join(SANDBOX_DIR, "services"), exist_ok=True)
    os.makedirs(os.path.join(SANDBOX_DIR, "scripts"), exist_ok=True)


def run_script(name: str, code: str, language: str = "python") -> str:
    init()
    ext = {"python": ".py", "bash": ".sh", "node": ".js"}.get(language, ".py")
    script_path = os.path.join(SANDBOX_DIR, "scripts", f"{name}{ext}")

    with open(script_path, "w") as f:
        f.write(code)
    os.chmod(script_path, 0o755)

    cmd = {"python": ["python", script_path],
           "bash": ["bash", script_path],
           "node": ["node", script_path]}.get(language, ["python", script_path])

    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120,
                                cwd=SANDBOX_DIR)
        output = result.stdout + result.stderr
        return output[:4000] if output else "(no output)"
    except subprocess.TimeoutExpired:
        return "[TIMEOUT after 120s]"
    except Exception as e:
        return f"[ERROR: {e}]"


def start_service(name: str, command: str) -> str:
    init()
    _load_processes()

    log_path = os.path.join(SANDBOX_DIR, "services", f"{name}.log")
    log_file = open(log_path, "a")

    try:
        proc = subprocess.Popen(
            command, shell=True, stdout=log_file, stderr=log_file,
            cwd=SANDBOX_DIR, start_new_session=True
        )
        _save_process(name, proc.pid, command)
        return f"service '{name}' started (pid={proc.pid}, log={log_path})"
    except Exception as e:
        return f"[ERROR starting service: {e}]"


def stop_service(name: str) -> str:
    processes = _load_processes()
    if name not in processes:
        return f"[no service named '{name}']"

    pid = processes[name]["pid"]
    try:
        os.kill(pid, signal.SIGTERM)
        del processes[name]
        _save_processes(processes)
        return f"service '{name}' stopped (pid={pid})"
    except ProcessLookupError:
        del processes[name]
        _save_processes(processes)
        return f"service '{name}' was already dead, cleaned up"
    except Exception as e:
        return f"[ERROR stopping service: {e}]"


def list_services() -> str:
    processes = _load_processes()
    if not processes:
        return "no running services"
    lines = []
    for name, info in processes.items():
        alive = _is_alive(info["pid"])
        status = "running" if alive else "dead"
        lines.append(f"- {name}: pid={info['pid']} [{status}] cmd={info['command']}")
    return "\n".join(lines)


def read_service_log(name: str, tail: int = 50) -> str:
    log_path = os.path.join(SANDBOX_DIR, "services", f"{name}.log")
    if not os.path.exists(log_path):
        return f"[no log for service '{name}']"
    with open(log_path) as f:
        lines = f.readlines()
    return "".join(lines[-tail:])[:4000]


def install_package(package: str) -> str:
    try:
        result = subprocess.run(
            ["pip", "install", package],
            capture_output=True, text=True, timeout=120
        )
        return (result.stdout + result.stderr)[:2000]
    except Exception as e:
        return f"[INSTALL ERROR: {e}]"


def create_project(name: str, files: dict[str, str]) -> str:
    init()
    project_dir = os.path.join(SANDBOX_DIR, "projects", name)
    os.makedirs(project_dir, exist_ok=True)

    created = []
    for filename, content in files.items():
        filepath = os.path.join(project_dir, filename)
        os.makedirs(os.path.dirname(filepath), exist_ok=True)
        with open(filepath, "w") as f:
            f.write(content)
        created.append(filename)

    return f"project '{name}' created at {project_dir} with files: {', '.join(created)}"


def _is_alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
        return True
    except (ProcessLookupError, PermissionError):
        return False


def _load_processes() -> dict:
    if not os.path.exists(PROCESSES_FILE):
        return {}
    try:
        with open(PROCESSES_FILE) as f:
            return json.loads(f.read())
    except Exception:
        return {}


def _save_process(name: str, pid: int, command: str):
    processes = _load_processes()
    processes[name] = {"pid": pid, "command": command, "started": time.time()}
    _save_processes(processes)


def _save_processes(processes: dict):
    with open(PROCESSES_FILE, "w") as f:
        f.write(json.dumps(processes))

File: lib/adam/tools/sandbox.ex
defmodule Adam.Tools.Sandbox do
  @sandbox_dir "/app/sandbox"

  def run_script(%{"code" => code, "language" => lang}) when is_binary(code) and is_binary(lang) do
    File.mkdir_p!(@sandbox_dir)
    {ext, cmd} = runner_for(lang)
    script_path = Path.join(@sandbox_dir, "script#{ext}")
    File.write!(script_path, code)

    try do
      case System.cmd(cmd, [script_path], cd: @sandbox_dir, stderr_to_stdout: true) do
        {output, 0} -> String.slice(output, 0, 2000)
        {output, code} -> "[EXIT #{code}] #{String.slice(output, 0, 2000)}"
      end
    rescue
      _ -> "[TIMEOUT]"
    after
      File.rm(script_path)
    end
  end

  def run_script(args), do: "[ERROR: sandbox_run requires 'code' and 'language' strings, got #{inspect(args)}]"

  def start_service(%{"name" => name, "command" => command}) when is_binary(name) and is_binary(command) do
    File.mkdir_p!(@sandbox_dir)
    log_path = Path.join(@sandbox_dir, "#{name}.log")
    pid_path = Path.join(@sandbox_dir, "#{name}.pid")

    System.cmd("sh", ["-c", "nohup #{command} > #{log_path} 2>&1 & echo $!"],
      cd: @sandbox_dir,
      stderr_to_stdout: true
    )
    |> case do
      {pid_str, 0} ->
        File.write!(pid_path, String.trim(pid_str))
        "started service '#{name}' (pid: #{String.trim(pid_str)})"

      {output, _} ->
        "[ERROR starting #{name}]: #{output}"
    end
  end

  def start_service(args), do: "[ERROR: sandbox_service_start requires 'name' and 'command', got #{inspect(args)}]"

  def stop_service(%{"name" => name}) when is_binary(name) do
    pid_path = Path.join(@sandbox_dir, "#{name}.pid")

    if File.exists?(pid_path) do
      pid = File.read!(pid_path) |> String.trim()
      System.cmd("kill", [pid], stderr_to_stdout: true)
      File.rm(pid_path)
      "stopped service '#{name}'"
    else
      "[ERROR: service '#{name}' not found]"
    end
  end

  def stop_service(args), do: "[ERROR: sandbox_service_stop requires 'name', got #{inspect(args)}]"

  def list_services do
    File.mkdir_p!(@sandbox_dir)

    Path.wildcard(Path.join(@sandbox_dir, "*.pid"))
    |> Enum.flat_map(fn path ->
      name = Path.basename(path, ".pid")
      case File.read(path) do
        {:ok, raw} -> ["#{name} (pid: #{String.trim(raw)})"]
        _ -> []
      end
    end)
    |> case do
      [] -> "no running services"
      services -> Enum.join(services, "\n")
    end
  end

  def read_log(%{"name" => name}) when is_binary(name) do
    log_path = Path.join(@sandbox_dir, "#{name}.log")

    if File.exists?(log_path) do
      File.read!(log_path) |> String.slice(-2000, 2000)
    else
      "[ERROR: no log for '#{name}']"
    end
  end

  def read_log(args), do: "[ERROR: sandbox_log requires 'name', got #{inspect(args)}]"

  def install_package(%{"package" => package, "manager" => manager}) when is_binary(package) and is_binary(manager) do
    cmd =
      case manager do
        "pip" -> "pip install #{package}"
        "npm" -> "npm install -g #{package}"
        "apt" -> "apt-get install -y #{package}"
        _ -> "echo 'unknown package manager: #{manager}'"
      end

    case System.cmd("sh", ["-c", cmd], cd: @sandbox_dir, stderr_to_stdout: true) do
      {output, 0} -> "installed #{package}\n#{String.slice(output, 0, 500)}"
      {output, _} -> "[INSTALL ERROR] #{String.slice(output, 0, 1000)}"
    end
  end

  def install_package(args), do: "[ERROR: sandbox_install requires 'package' and 'manager', got #{inspect(args)}]"

  def create_project(%{"name" => name, "template" => template}) when is_binary(name) and is_binary(template) do
    project_dir = Path.join(@sandbox_dir, name)
    File.mkdir_p!(project_dir)

    case template do
      "python" ->
        File.write!(Path.join(project_dir, "main.py"), "#!/usr/bin/env python3\n\n")
        File.write!(Path.join(project_dir, "requirements.txt"), "")

      "node" ->
        System.cmd("npm", ["init", "-y"], cd: project_dir, stderr_to_stdout: true)

      "elixir" ->
        System.cmd("mix", ["new", name, "--sup"], cd: @sandbox_dir, stderr_to_stdout: true)

      _ ->
        File.write!(Path.join(project_dir, "README.md"), "# #{name}\n")
    end

    "created project '#{name}' (#{template}) at sandbox/#{name}"
  end

  def create_project(args), do: "[ERROR: sandbox_project requires 'name' and 'template', got #{inspect(args)}]"

  defp runner_for("python"), do: {".py", "python3"}
  defp runner_for("node"), do: {".js", "node"}
  defp runner_for("bash"), do: {".sh", "bash"}
  defp runner_for("elixir"), do: {".exs", "elixir"}
  defp runner_for(_), do: {".sh", "bash"}
end
